common/cryptfs.c

```diff
@@ -194,6 +281,114 @@ load_crypto_mapping_table(const char *real_blk_name, const char *master_key_asci
 	}
 }
 
+// This function does:
+//		Create an integrity block device with IOCTL(DM_DEV_CREATE)
+//		Next load the mapping table of this device with IOCTL(DM_TABLE_LOAD)
+//		Lastly resume the device with IOCTL(DM_DEV_SUSPEND)
+//		Only then the crypto device can be mounted ontop of the integrity device
+//
+//	This general functionality is adapted from [1,2] and strace of the dmsetup setup
+//	When executing this functionality on its on (in a separate .c file), on a host system it works
+//	Verified by:
+//		user@host:~/$ dmsetup ls --tree
+//			00000000-0000-0000-0000-000000000000-etc (252:1)
+//			 └─00000000-0000-0000-0000-000000000000-etc-integrity (252:0)
+//			    └─ (7:3)
+
+//
+//	And:
+//      user@host:~/$ sudo dmsetup table
+//            dmsetup table
+//				0[...]0-etc-integrity: 0 1677721 integrity 7:3 0 32 J 5 journal_sectors:16368 interleave_sectors:32768 buffer_sectors:128 journal_watermark:50 commit_time:10000
+//				0[...]0-etc: 0 1677721 crypt capi:aegis128-random 00000000000000000000000000000000 0 252:0 0 1 integrity:32:aead
+//
+// [1] https://www.kernel.org/doc/html/latest/admin-guide/device-mapper/dm-integrity.html
+// [2] https://wiki.gentoo.org/wiki/Device-mapper#Integrity
+static int
+create_integrity_blk_dev(const char *real_blk_name, const char *name, const unsigned long fs_size)
+{
+	int fd_mapper;
+	int ioctl_ret;
+	int load_count = -1;
+	char create_buffer[DM_INTEGRITY_BUF_SIZE];
+	struct dm_ioctl *create_io;
+	int create_counter;
+
+	//open device mapper
+	if ((fd_mapper = open(DEV_MAPPER, O_RDWR)) < 0) {
+		ERROR("Cannot open device-mapper");
+	}
+
+	// Create blk device
+	//initialize create_io struct
+	create_io = (struct dm_ioctl *)create_buffer;
+	ioctl_init(create_io, DM_INTEGRITY_BUF_SIZE, name, 0);
+
+	for (create_counter = 0; create_counter < TABLE_LOAD_RETRIES; create_counter++) {
+		ioctl_ret = dm_ioctl(fd_mapper, DM_DEV_CREATE, create_io);
+		if (ioctl_ret != 0) {
+			ERROR("Could not create block device: ioctl(DM_DEV_CREATE: '%lu'|int: %d ) ret: %d, errno: %d",
+			      DM_DEV_CREATE, (int)DM_DEV_CREATE, ioctl_ret, errno);
+		} else {
+			DEBUG("Creating block device worked!");
+			break;
+		}
+
+		usleep(500000);
+	}
+
+	if (create_counter >= TABLE_LOAD_RETRIES) {
+		ERROR("Failed to create block device after %d tries", create_counter);
```

Comment:
ERROR_ERRNO

btw, is this retry really necessary in every case? shouldn't the errno be check to values which are temporarily erros such as EBUSY ?.

Further, since i read this retry loop 3 times now, it would make sense to put this directly in the dm_ioctl wrapper

Reply:
Not every call to dm_ioctl uses this loop (e.g. dm_ioctl(fd, DM_DEV_STATUS, io)), so I think it would not be suitable in the wrapper.